File: flash_api.go
package mineru

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

// TODO(release): 上线前换回 https://mineru.net/api/v1/agent
const defaultFlashBaseURL = "https://staging.mineru.org.cn/api/v1/agent"

// flashApiClient is the low-level HTTP wrapper for the Flash (agent) API.
// It never sends an Authorization header.
type flashApiClient struct {
	httpClient *http.Client
	baseURL    string
	source     string
}

func (a *flashApiClient) post(ctx context.Context, path string, payload any) (json.RawMessage, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, a.baseURL+path, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if a.source != "" {
		req.Header.Set("source", a.source)
	}
	return a.do(req)
}

func (a *flashApiClient) get(ctx context.Context, path string) (json.RawMessage, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, a.baseURL+path, nil)
	if err != nil {
		return nil, err
	}
	return a.do(req)
}

func (a *flashApiClient) putFile(ctx context.Context, url string, data []byte) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodPut, url, bytes.NewReader(data))
	if err != nil {
		return err
	}
	resp, err := a.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("upload failed: HTTP %d: %s", resp.StatusCode, body)
	}
	return nil
}

func (a *flashApiClient) downloadText(ctx context.Context, url string) (string, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return "", err
	}
	resp, err := a.httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("download failed: HTTP %d", resp.StatusCode)
	}
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("read response: %w", err)
	}
	return string(data), nil
}

func (a *flashApiClient) do(req *http.Request) (json.RawMessage, error) {
	resp, err := a.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}
	if resp.StatusCode == 429 {
		return nil, &ParamError{APIError{Code: "RATE_LIMITED", Message: "flash API rate limit exceeded; try again later"}}
	}
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, body)
	}
	var ar apiResponse
	if err := json.Unmarshal(body, &ar); err != nil {
		return nil, fmt.Errorf("unmarshal response: %w", err)
	}
	codeStr := codeToString(ar.Code)
	if codeStr != "0" {
		return nil, errorForCode(codeStr, ar.Msg, ar.TraceID)
	}
	return ar.Data, nil
}

File: src/mineru/_flash_api.py
"""Low-level HTTP wrapper for the Flash (agent) API."""

from __future__ import annotations

from typing import Any

import httpx

from .exceptions import raise_for_code

# TODO(release): 上线前换回 https://mineru.net/api/v1/agent
DEFAULT_FLASH_BASE_URL = "https://staging.mineru.org.cn/api/v1/agent"

_TIMEOUT = httpx.Timeout(30.0, read=120.0)


class FlashApiClient:
    """No-auth HTTP client for the Flash API."""

    def __init__(self, base_url: str = DEFAULT_FLASH_BASE_URL) -> None:
        self._client = httpx.Client(
            base_url=base_url,
            headers={"Content-Type": "application/json"},
            timeout=_TIMEOUT,
        )

    def close(self) -> None:
        self._client.close()

    def post(self, path: str, json: dict[str, Any]) -> dict[str, Any]:
        resp = self._client.post(path, json=json)
        return self._handle(resp)

    def get(self, path: str) -> dict[str, Any]:
        resp = self._client.get(path)
        return self._handle(resp)

    def put_file(self, url: str, data: bytes) -> None:
        resp = httpx.put(url, content=data, timeout=_TIMEOUT)
        resp.raise_for_status()

    def download_text(self, url: str) -> str:
        resp = httpx.get(url, timeout=httpx.Timeout(30.0, read=300.0), follow_redirects=True)
        resp.raise_for_status()
        return resp.text

    @staticmethod
    def _handle(resp: httpx.Response) -> dict[str, Any]:
        if resp.status_code == 429:
            raise_for_code("RATE_LIMITED", "flash API rate limit exceeded; try again later")
        resp.raise_for_status()
        body: dict[str, Any] = resp.json()
        code = body.get("code", 0)
        if code != 0:
            raise_for_code(code, body.get("msg", "unknown error"), body.get("trace_id", ""))
        return body
